File: sandbox/guest/sync-agent/dataplane.go
package main

import (
	"bufio"
	"encoding/json"
	"log"
	"net"
	"strconv"
	"sync"
	"sync/atomic"
	"time"
)

// dataPlaneCfg is advertised by the host over the console (HELLO / hello-ACK
// payloads). The guest dials ip:port through virtio-net; the host's wisp
// relay bridges that stream in-process. token gates the channel per boot.
type dataPlaneCfg struct {
	IP    string `json:"ip"`
	Port  int    `json:"port"`
	Token string `json:"token"`
}

// DataPlane maintains the TCP sync channel: dial with retry, token HELLO,
// frame session sharing SyncState with the console session, liveness pings,
// re-dial on drop or on a new advert (host restart ⇒ new token).
type DataPlane struct {
	mu     sync.Mutex
	root   string
	ss     *SyncState
	cfg    *dataPlaneCfg
	gen    int
	conn   net.Conn
	sender *Sender
}

func NewDataPlane(root string, ss *SyncState) *DataPlane {
	return &DataPlane{root: root, ss: ss}
}

// Update adopts a (possibly new) advert. A changed advert supersedes the
// current session: the old connection is closed and a fresh dial loop starts.
func (dp *DataPlane) Update(cfg dataPlaneCfg) {
	dp.mu.Lock()
	if dp.cfg != nil && *dp.cfg == cfg {
		dp.mu.Unlock()
		return
	}
	dp.cfg = &cfg
	dp.gen++
	gen := dp.gen
	if dp.conn != nil {
		dp.conn.Close()
		dp.conn = nil
	}
	dp.mu.Unlock()
	log.Printf("data plane: advert %s:%d (gen %d)", cfg.IP, cfg.Port, gen)
	go dp.loop(cfg, gen)
}

func (dp *DataPlane) Shutdown() {
	dp.mu.Lock()
	dp.gen++ // invalidate all loops
	if dp.conn != nil {
		dp.conn.Close()
		dp.conn = nil
	}
	dp.mu.Unlock()
}

// Sender returns the active data-channel sender, or nil when disconnected.
func (dp *DataPlane) Sender() *Sender {
	dp.mu.Lock()
	defer dp.mu.Unlock()
	return dp.sender
}

func (dp *DataPlane) stale(gen int) bool {
	dp.mu.Lock()
	defer dp.mu.Unlock()
	return dp.gen != gen
}

func (dp *DataPlane) loop(cfg dataPlaneCfg, gen int) {
	addr := net.JoinHostPort(cfg.IP, strconv.Itoa(cfg.Port))
	for {
		if dp.stale(gen) {
			return
		}
		conn, err := net.DialTimeout("tcp", addr, 5*time.Second)
		if err != nil {
			time.Sleep(2 * time.Second)
			continue
		}
		dp.mu.Lock()
		if dp.gen != gen {
			dp.mu.Unlock()
			conn.Close()
			return
		}
		dp.conn = conn
		dp.mu.Unlock()
		dp.session(conn, cfg, gen)
		dp.mu.Lock()
		if dp.conn == conn {
			dp.conn = nil
		}
		dp.mu.Unlock()
		time.Sleep(time.Second)
	}
}

func (dp *DataPlane) session(conn net.Conn, cfg dataPlaneCfg, gen int) {
	defer conn.Close()
	fw := NewFrameWriter(conn)
	hello, _ := json.Marshal(map[string]any{
		"version": 1, "role": "guest", "channel": "data",
		"token": cfg.Token, "root": dp.root,
	})
	if _, err := fw.Send(TypeHello, hello); err != nil {
		return
	}
	recv := NewReceiverNoVerify(dp.root, fw, dp.ss)
	// disjoint xfer-id base: the console sender starts at 0, the host at
	// 0x80000000 — collisions would cross wires in per-xfer routing maps
	send := NewSenderWithBase(dp.root, fw, dp.ss, 0x40000000)

	dp.mu.Lock()
	dp.sender = send
	dp.mu.Unlock()
	defer func() {
		dp.mu.Lock()
		if dp.sender == send {
			dp.sender = nil
		}
		dp.mu.Unlock()
	}()
	log.Printf("data plane: connected to %s:%d", cfg.IP, cfg.Port)

	// liveness: ping every 15s; a snapshot restore leaves this TCP session
	// half-dead (host side is gone), so silence >45s forces a re-dial
	var lastRx atomic.Int64
	lastRx.Store(time.Now().UnixMilli())
	stop := make(chan struct{})
	defer close(stop)
	go func() {
		t := time.NewTicker(15 * time.Second)
		defer t.Stop()
		for {
			select {
			case <-t.C:
				fw.Send(TypePing, nil)
				if time.Now().UnixMilli()-lastRx.Load() > 45000 {
					log.Printf("data plane: no traffic for 45s — reconnecting")
					conn.Close()
					return
				}
			case <-stop:
				return
			}
		}
	}()

	r := bufio.NewReaderSize(conn, 256*1024)
	for {
		f, err := ReadFrame(r)
		if err != nil {
			log.Printf("data plane: session ended: %v", err)
			return
		}
		if dp.stale(gen) {
			return
		}
		lastRx.Store(time.Now().UnixMilli())
		switch f.Type {
		case TypeAck, TypeNak:
			send.HandleAck(f) // non-xfer acks (hello/ping replies) need nothing
		case TypeFilePut:
			recv.HandlePut(f)
		case TypeTreePut:
			recv.HandleTreePut(f)
		case TypeFileChunk:
			recv.HandleChunk(f)
		case TypeFileDel:
			recv.HandleDel(f)
		case TypePing:
			recv.ack(f.Seq, nil)
		case TypeManifest:
			recv.ack(f.Seq, nil) // manifests travel on the console today
		default:
			log.Printf("data plane: unknown frame type %d", f.Type)
		}
	}
}

File: sandbox/guest/sync-agent-rust/src/dataplane.rs

use std::io::BufReader;
use std::net::TcpStream;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Instant;

use crate::frame::{self, FrameWriter, TYPE_ACK, TYPE_HELLO, TYPE_NAK, TYPE_PING};
use crate::state::SyncState;
use crate::transfer::{new_receiver, new_sender};

#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, PartialEq, Eq)]
pub struct DataPlaneCfg {
    pub ip: String,
    pub port: i32,
    pub token: String,
}

struct DataPlaneInner {
    root: String,
    ss: Arc<SyncState>,
    cfg: Option<DataPlaneCfg>,
    gen: i32,
    sender: Option<Arc<crate::transfer::Sender>>,
}

pub struct DataPlane {
    inner: Arc<Mutex<DataPlaneInner>>,
}

impl DataPlane {
    pub fn new(root: &str, ss: Arc<SyncState>) -> Self {
        Self {
            inner: Arc::new(Mutex::new(DataPlaneInner {
                root: root.to_string(),
                ss,
                cfg: None,
                gen: 0,
                sender: None,
            })),
        }
    }

    pub fn update(&self, cfg: DataPlaneCfg) {
        let mut inner = self.inner.lock().unwrap();
        if inner.cfg.as_ref() == Some(&cfg) {
            return;
        }
        inner.cfg = Some(cfg.clone());
        inner.gen += 1;
        let gen = inner.gen;
        drop(inner);

        crate::slog!("data plane: advert {}:{} (gen {})", cfg.ip, cfg.port, gen);

        let inner_clone = self.inner.clone();
        thread::spawn(move || {
            dp_dial_loop(cfg, gen, inner_clone);
        });
    }

    #[allow(dead_code)]
    pub fn shutdown(&self) {
        let mut inner = self.inner.lock().unwrap();
        inner.gen += 1;
        inner.sender = None;
    }

    pub fn sender(&self) -> Option<Arc<crate::transfer::Sender>> {
        let inner = self.inner.lock().unwrap();
        inner.sender.clone()
    }

    #[allow(dead_code)]
    fn is_stale(&self, gen: i32) -> bool {
        let inner = self.inner.lock().unwrap();
        inner.gen != gen
    }
}

fn dp_dial_loop(cfg: DataPlaneCfg, gen: i32, inner: Arc<Mutex<DataPlaneInner>>) {
    let addr = format!("{}:{}", cfg.ip, cfg.port);
    loop {
        {
            let inn = inner.lock().unwrap();
            if inn.gen != gen {
                return;
            }
        }
        let conn = match TcpStream::connect_timeout(&addr.parse().unwrap(), std::time::Duration::from_secs(5)) {
            Ok(c) => c,
            Err(e) => {
                crate::slog!("data plane: dial {} failed: {}", addr, e);
                thread::sleep(std::time::Duration::from_secs(2));
                continue;
            }
        };

        {
            let inn = inner.lock().unwrap();
            if inn.gen != gen {
                conn.shutdown(std::net::Shutdown::Both).ok();
                return;
            }
        }

        dp_session(conn, cfg.clone(), gen, inner.clone());
        thread::sleep(std::time::Duration::from_secs(1));
    }
}

fn dp_session(conn: TcpStream, cfg: DataPlaneCfg, gen: i32, inner: Arc<Mutex<DataPlaneInner>>) {
    conn.set_nodelay(true).ok();
    let root;
    let ss: Arc<SyncState>;
    {
        let inn = inner.lock().unwrap();
        root = inn.root.clone();
        ss = inn.ss.clone();
    }

    let fw = Arc::new(FrameWriter::new(Box::new(conn.try_clone().unwrap())));

    let hello = serde_json::to_vec(&serde_json::json!({
        "version": 1,
        "role": "guest",
        "channel": "data",
        "token": cfg.token,
        "root": root,
    })).unwrap();
    if fw.send(TYPE_HELLO, &hello).is_err() {
        return;
    }

    let recv = Arc::new(new_receiver(&root, fw.clone(), ss.clone(), false));
    let send = Arc::new(new_sender(&root, fw.clone(), ss.clone(), 0x40000000));

    {
        let mut inn = inner.lock().unwrap();
        inn.sender = Some(send.clone());
    }

    crate::slog!("data plane: connected to {}:{:?}", cfg.ip, cfg.port);

    let last_rx_ts = Arc::new(AtomicI64::new(Instant::now().elapsed().as_millis() as i64));
    let last_rx_clone = last_rx_ts.clone();
    let fw_ping = fw.clone();
    let conn_for_close = conn.try_clone().unwrap();

    let _ping_handle = thread::spawn(move || {
        let start = Instant::now();
        loop {
            thread::sleep(std::time::Duration::from_secs(15));
            let _ = fw_ping.send(TYPE_PING, &[]);
            let elapsed = start.elapsed().as_millis() as i64;
            let last = last_rx_clone.load(Ordering::SeqCst);
            if elapsed - last > 45000 {
                crate::slog!("data plane: no traffic for 45s — reconnecting");
                conn_for_close.shutdown(std::net::Shutdown::Both).ok();
                return;
            }
        }
    });

    let mut reader = BufReader::with_capacity(256 * 1024, conn.try_clone().unwrap());
    loop {
        {
            let inn = inner.lock().unwrap();
            if inn.gen != gen {
                return;
            }
        }
        let f = match frame::read_frame(&mut reader) {
            Ok(f) => f,
            Err(e) => {
                crate::slog!("data plane: session ended: {}", e);
                return;
            }
        };
        last_rx_ts.store(Instant::now().elapsed().as_millis() as i64, Ordering::SeqCst);

        match f.typ {
            TYPE_ACK | TYPE_NAK => {
                // non-xfer acks need nothing
            }
            crate::frame::TYPE_FILE_PUT => {
                recv.handle_put(&f);
            }
            crate::frame::TYPE_TREE_PUT => {
                recv.handle_tree_put(&f);
            }
            crate::frame::TYPE_FILE_CHUNK => {
                recv.handle_chunk(&f);
            }
            crate::frame::TYPE_FILE_DEL => {
                recv.handle_del(&f);
            }
            TYPE_PING => {
                let ack_payload = serde_json::to_vec(&serde_json::json!({"ack": f.seq})).unwrap();
                let _ = fw.send(TYPE_ACK, &ack_payload);
            }
            crate::frame::TYPE_MANIFEST => {
                let ack_payload = serde_json::to_vec(&serde_json::json!({"ack": f.seq})).unwrap();
                let _ = fw.send(TYPE_ACK, &ack_payload);
            }
            _ => {
                crate::slog!("data plane: unknown frame type {}", f.typ);
            }
        }
    }
}
